nodepool/elements/virt-customize/create-debian-vm-image.sh

```diff
@@ -0,0 +1,178 @@
+#!/bin/bash
+
+set -e
+set -x
+
+fatal_error() {
+    echo "$*" 1>&2
+    exit 1
+}
+
+do_cleanup() {
+    ret=$?
+    trap - 0
+    set +e
+
+    rm -rf "$mdir/"dev/*
+    umount "$mdir/"proc
+    umount "$mdir"
+
+    losetup -d $dev
+    try=5
+    while [ $try -gt 0 ] && ! kpartx -d "$disk"; do
+	sleep 1
+	try=$(($try - 1))
+    done
+    losetup -d $disk
+    rmdir "$mdir"
+}
+
+check_binary() {
+    type -p $1 || fatal_error "$1 is missing"
+}
+
+do_chroot() {
+    local chdir="$1"
+    shift
+    PATH=/bin/:/sbin:$PATH LANG=C LC_ALL=C LC_CTYPE=C LANGUAGE=C chroot "$chdir" "$@"
+}
+
+if [ $(whoami) != root ]; then
+    fatal_error "restart as root like sudo $0 $*"
+fi
+
+dir=$(cd $(dirname $0); pwd)
+
+if [ $# != 1 -a $# != 2 ]; then
+    fatal_error "Usage: $0 <distro name> [custom script]"
+fi
+
+distro=$1
+img=$distro.raw
+custom="$2"
+
+# compress image in qcow2
+comp_opt=-c
+# no compression of the image in qcow2
+#comp_opt=
+
+export DEBOOTSTRAP_DIR=$dir/debootstrap
+
+if [ -r $dir/config ]; then
+    . $dir/config
+fi
+
+FSTYPE=${FSTYPE:=ext2}
+
+modprobe loop
+
+for cmd in git du dd parted chroot kpartx qemu-img losetup mkfs.$FSTYPE rsync; do
+    check_binary $cmd
+done
+
+if [ ! -d $DEBOOTSTRAP_DIR/.git ]; then
+    rm -rf $DEBOOTSTRAP_DIR
+    git clone https://salsa.debian.org/installer-team/debootstrap.git $DEBOOTSTRAP_DIR
+else
+    pushd $DEBOOTSTRAP_DIR
+    git pull
+    popd
+fi
+
+rm -rf $distro
+mkdir -p $distro
+
+$dir/debootstrap/debootstrap --include linux-image-amd64,grub-pc,bash --arch amd64 $distro $distro http://deb.debian.org/debian
+
+# some scriptlets need /proc
+mount --bind /proc $distro/proc
+
+# run the custom script if any
+if [ -x "$custom" ]; then
+    set +e
+    "$custom" $distro
+    rc=$?
+    set -e
+fi
+
+umount $distro/proc
+
+if [ $rc != 0 ]; then
+    exit $rc
+fi
+
+# cleanup distro
+do_chroot $distro apt-get autoremove -y
+
+rm -f $distro/var/cache/apt/archives/*.deb
+
+# compute size of the directory
+size=$(du -s -BM "$distro" | cut -f1 | sed -s 's/.$//')
+
+# add 30% to be sure that metadata from the filesystem fit
+size=$(($size * 130 / 100))
+
+# Create the image file
+rm -f $img
+dd if=/dev/zero of=$img count=0 bs=1M seek=$size
+
+disk=$(losetup --show --find "$img")
+parted -s "$disk" mklabel msdos
+parted -s "$disk" mkpart primary $FSTYPE 32k '100%'
+parted "$disk" set 1 boot on
+
+part=/dev/mapper/$(kpartx -av $disk|cut -f3 -d' ')
+
+try=5
+while [ $try -gt 0 -a ! -b $part ]; do
+    sleep 1
+    try=$(($try - 1))
+done
+
+# create filesystem
+rsync -aX --delete-before --exclude=shm /dev/ $distro/dev/
```

Comment:
I personally used mount -o bind for /dev and /dev/pts to keep the final image with a empty /dev

Reply:
yes but we need the extra devices created in between